js/imageeditor/shared/tools/brush.ts

```diff
@@ -0,0 +1,246 @@
+import {
+	Container,
+	Graphics,
+	Color,
+	type ColorSource,
+	type IRenderer,
+	RenderTexture,
+	Sprite
+} from "pixi.js";
+
+import type { LayerScene } from "../layers/utils";
+import { type Command } from "../utils/commands";
+import { make_graphics } from "../utils/pixi";
+
+export interface DrawCommand extends Command {
+	/**
+	 * sets the initial state to begin drawing
+	 * @param options the options for drawing
+	 * @returns
+	 */
+	start: (options: DrawOptions) => void;
+	/**
+	 * continues drawing, smoothly interpolating between points where necessary
+	 * @param options the options for drawing
+	 * @returns
+	 */
+	continue: (options: Points) => void;
+	/**
+	 * stops drawing
+	 * @returns
+	 */
+	stop: () => void;
+	/**
+	 * Whether or not the user is currently drawing.
+	 */
+	drawing: boolean;
+}
+
+/**
+ * Coordinates for a draw or erase path.
+ */
+interface Points {
+	/**
+	 * The x coordinate of the path.
+	 */
+	x: number;
+	/**
+	 * The y coordinate of the path.
+	 */
+	y: number;
+}
+
+/**
+ * Options for drawing a path.
+ */
+interface DrawOptions extends Points {
+	/**
+	 * The size of the brush.
+	 */
+	size: number;
+	/**
+	 * The color of the brush.
+	 */
+	color?: ColorSource;
+	/**
+	 * The opacity of the brush.
+	 */
+	opacity: number;
+	/**
+	 * Whether or not to set the initial texture.
+	 */
+	set_initial_texture?: boolean;
+}
+
+/**
+ * Draws a circle on the given graphics object.
+ * @param graphics the graphics object to draw on
+ * @param x the x coordinate of the circle
+ * @param y the y coordinate of the circle
+ * @param brush_color the color of the circle
+ * @param brush_size the radius of the circle
+ */
+function drawCircle(
+	graphics: Graphics,
+	x: number,
+	y: number,
+	brush_color: ColorSource = new Color("black"),
+	brush_size: number
+): void {
+	const color = new Color(brush_color);
+	graphics.beginFill(color);
+	graphics.drawCircle(x, y, brush_size);
+	graphics.endFill();
+}
+
+/**
+ * Interpolates between two points.
+ * @param point1 the first point
+ * @param point2 the second point
+ * @returns an array of points between the two points
+ */
+function interpolate(
+	point1: { x: number; y: number },
+	point2: { x: number; y: number }
+): Points[] {
+	let points: Points[] = [];
+	const dx = point2.x - point1.x;
+	const dy = point2.y - point1.y;
+	const distance = Math.sqrt(dx * dx + dy * dy);
+	const steps = Math.ceil(distance / 2);
+	const stepX = dx / steps;
+	const stepY = dy / steps;
+
+	for (let j = 0; j < steps; j++) {
+		const x = point1.x + j * stepX;
+		const y = point1.y + j * stepY;
+		points.push({ x, y });
+	}
+
+	return points;
+}
```

Comment:
This seems to cause performance issues with a large brush size. Not 100% sure why but a simply limiting per frame in the handler might be enough.